shell/testing/tester_main.cc

```diff
@@ -44,25 +44,28 @@ class ScriptCompletionTaskObserver {
   }
 
   void DidProcessTask() {
-    last_error_ = shell_.GetUIIsolateLastError();
-    if (shell_.EngineHasLivePorts()) {
-      // The UI isolate still has live ports and is running. Nothing to do
-      // just yet.
-      return;
-    }
-
-    if (run_forever_) {
-      // We need this script to run forever. We have already recorded the last
-      // error. Keep going.
-      return;
-    }
-
-    if (!has_terminated) {
-      // Only try to terminate the loop once.
-      has_terminated = true;
-      main_task_runner_->PostTask(
-          []() { fml::MessageLoop::GetCurrent().Terminate(); });
-    }
+    // We are installed on the UI thread so we need to move to the platform task
+    // runner
+    fml::TaskRunner::RunNowOrPostTask(main_task_runner_, [this]() {
```

Comment:
DidProcessTask is going to get called extremely frequently and the callback is meant to be extremely fast. So the cross thread latching should be avoided.

Also, The `EngineHasLivePorts` call needs to happen on the UI Task runner because we need to enter isolate scope to tell if the isolate has live ports. So, the script completion observer should run on the UI task runner. However, when the observer determines that there are no more live ports, the main task runner needs to be terminated. This is how we exit the main thread.

The following pseudo-code should guarantee that the callback is fast in the usual case, is thread safe while making the call to check if the engine has live ports, and terminate the right task runner.

1. Add script completion observer to the UI Task runner (already done correctly).
2. Make the EngineHasLivePorts call directly in the `DidProcessTask` callback (unlike the main_task_runner_ in the current state of the patch).
3. If there are no live ports, post a task to the main task runner that terminates that task runner.

Reply:
Unfortunately this won't work without changing the thread requirements of Shell. I figured because this is test-only code it's not on a perf-critical path and the cross-thread latching (in the `EngineHasLivePorts` call) would be acceptable.